README.md
# elitunes
A stable but hastily scraped together summer project

#Notable files:

changelog.txt - change tracker

index.tpl - main webpage

musicdb.json - music database (excluded from repo)

musicdb2.py - all the functions

player.css - stylesheet

player.js - webpage functions

webplayer.py - main server script

The rest are outdated/test files or unix commands

#--Currently on hiatus--


changelog.txt
v0.9 --Reticulating Splines--
    July 25, 2016
Added playlist function
1. Pulls list+data from current query displayed
2. Highlights current song
3. Previous/Next buttons functional
4. Also controllable via arrow keys
5. Next song starts automatically
Added Play Next button function to add songs
Added song file/DB updating via user interface
Added logging for easier debugging
Client database updated silently on the fly with
    no page reload
Added file lock per song while updating
Added hash history storage for file changes
Added search function by clicking footer details
Added album art easter egg
    Removable via escape key
Added scrollto function for elements out of view
Cleaned up webpage scripting (couple thousand
    loose global variables)
Recently added now sorted by date added
Lots more sorting and import fixes
Imported half my personal library (~5000 songs)
    without any noticable hits to performance or
    page size (remains around 1mb)
Database is 1/3 size compared to iTunes'
Multiple concurrent users handled safely and
    sessions can persist through server resets
Mobile browsers currently crash the entire server
    when closing the window

#################################################

v0.8 --iTunes Understudy--
    July 17, 2016
Added support for album art
Added loading screen
Added smooth scrolling alphabet selector for
    artist panel
Added options bar for songs via right click
Added Download and details buttons to options
Details table beefed up and editable by client
Drastically improved performance issues:
1. Live updating now entirely handled by JS
    Not so hacky anymore, objects are added and
    removed dynamically, more lightweight
2. JSONs converted to JS object on server
3. Defaults to 250 Recently added songs instead
    of entire library so there aren't 5000
    objects on the page immediately
4. Database *LITE* is pushed to client on first
    pageload
5. Filepaths sent to client so songs play
    immediately, then details are fetched
6. Page footprint chopped down to 1/3 of previous
    ~3mb -> 1mb
Fixed some song spacing issues in some browsers
Fixed non-ascii characters slipping through

#################################################

v0.7 --Control Rollout--
    July 16, 2016
Added player progress bar w seeking and time
Added pause/play button
Added currently playing footer
Added spacebar to toggle playback
Added volume slider
Added download song button
Added song details table
Added YouTube external search button
Cleaned up artists panel
Websocket will now process different arguments
1. playlists
2. song tags
3. returns json for javascript to parse
Fixed formatting so buttons don't break line
Fixed file save issue that only effected Jerome

#################################################

v0.6 --Vanish DatPiff n Rack'em Brackets--
    July 10, 2016
Added naming sieve
1. Moves features from artist tag to title
2. Formats (feat. ) and [prod ]
3. Strips DatPiff tags
4. Can guess artist/title from filename
5. Preserves comma credits in respect for Kanye
Updated color scheme
Source code reorganized with some comments
Updating tags is now possible serverside, again

#################################################

v0.5 --LIVE--
    July 8, 2016
Added working real-time search bar
Added sort by artist/album by artist default
Added leftside artist buttons
Added hacky method for live updating webpage
Websocket is kept alive now
Solved gevent python3 compatibility bug

#################################################

v0.4 --Powered By 3 Americas--
    June 27, 2016
Started serving on a websocket via gevent, though
    it behaves suspiciously
Added search bar
Added live updating Python search function piped
    through the websocket while studying the
    Ballmer Peak and Jerome racked a smooth drum
    on a track
Not until later would I find that it crashed the
    site quickly and consistently

#################################################

v0.3 --Rewrite with C-Types--
    June 21, 2016
Song tags now read via pytaglib instead of eyed3
    to support more than solely mp3 files
Revamped import functionality

#################################################

v0.2 --Straight from Michigan--
    May 2016
Added import function
Song tags now read and write via eyed3 instead of
    EasyTag which cannot update tags

#################################################

v0.1 --Pop Tags--
    January 2016
JSON database created
Divvied up CSS
Added footer to hold playback controls
Added artist pane structured with alphabet
Button labels pretty print now (showed URL before)
Song tags now read via EasyTag
Buttons show "Artist - Title"
Branded as "EliTunes"


test.py
import logging
root_dir = '/media/eli/shorty/music/'
static_dir='static/'
cfg_dir = root_dir + 'cfg/'
logging.basicConfig(filename=str(cfg_dir + "debug.log"), level=logging.DEBUG, format='%(asctime)s %(message)s', datefmt='%m/%d/%Y %I:%M:%S %p')
logging.info("TEST1")

while True:
    num = 1


musicdb.py
import json
import os
import eyed3
from tinytag import TinyTag
from shutil import copy2
from urllib.parse import quote
from hashlib import sha1
"""
tag.album         # album as string
tag.albumartist   # album artist as string
tag.artist        # artist name as string
tag.audio_offset  # number of bytes before audio data begins
tag.bitrate       # bitrate in kBits/s
tag.disc          # disc number
tag.disc_total    # the total number of discs
tag.duration      # duration of the song in seconds
tag.filesize      # file size in bytes
tag.genre         # genre as string
tag.samplerate    # samples per second
tag.title         # title of the song
tag.track         # track number as string
tag.track_total   # total number of tracks as string
tag.year          # year or data as string

tag = TinyTag.get('/some/music.mp3', image=True)
tag.get_image()
"""

fields = ('album', 'albumartist', 'artist', 'title', 'genre', 'track', 'year', 'plays', 'img', 'bitrate', 'lyrics', 'duration', 'uid', 'size', 'sorted')
fields_mod = ('album', 'albumartist', 'artist', 'title', 'genre', 'track', 'year', 'plays', 'img', 'lyrics')

root_dir = '/media/eli/shorty/music/'
backup_dir = root_dir + 'cfg/backup/'
library_dir = root_dir + 'library/'
import_dir = root_dir + 'import/'
source_dir = '/home/eli/Dropbox/code/player/source/'
asset_dir = source_dir + 'assets'
index_tpl = source_dir + 'index.tpl'
db_file = source_dir + 'musicdb.json'




def buildDB(contents): #takes dictionary and dumps it
    print(contents)
    with open(db_file, 'w', encoding='utf-8') as outfile:
        json.dump(contents, outfile, sort_keys = True, indent = 4, ensure_ascii = False)

def addToLibrary(song, artist, title): #Renames file, copies, and returns new file name
    if artist[0].isalpha() == True:
        first_let = artist[0].lower()
    else:
        first_let = '0'
    artist_dir = library_dir + first_let + os.sep
    filename = artist + '_' + title
    fileext = os.path.splitext(song)[1]
    dup_count = 1
    while os.path.exists(artist_dir + filename + fileext) == True:
        filename = artist + '_' + title + '_' + str(dup_count)
        dup_count += 1
    copy2((root_dir + song), (artist_dir + filename + fileext))
    return (filename + fileext)

def getImports(): #returns list of only files in /music
    root_path = os.listdir(root_dir)
    root_list = []
    for i in root_path:
        #name = str(i)
        if os.path.isdir(root_dir + i):
            continue
        else:
            root_list.append(i)
    return root_list

def getTag(song_path): #returns all matching fields
    tag = TinyTag.get(song_path)
    tag_fields = tag.__dict__
    details = {}
    for item in tag_fields:
        if item in fields and tag_fields[item] != None:
            details[item] = tag_fields[item]
    details['duration'] = "{:.2f}".format(details.get('duration'))
    if details.get('bitrate') > 5000:
        details.pop('bitrate')
    return details

def openDB(): #returns contents of musicdb
    with open(db_file, 'r', encoding='utf-8') as db_json:
        db = json.load(db_json)
        db_json.close()
        return db

def updateTag(song_path, tag_dict):
    song_file = eyed3.load(song_path)
    song_file.initTag(version=(2, 3, 0))
    if tag_dict.get('title') and tag_dict.get('title') != None:
        song_file.tag.title = tag_dict.get('title')
    if tag_dict.get('albumartist') and tag_dict.get('albumartist') != None:
        song_file.tag.album_artist = tag_dict.get('albumartist')
    if tag_dict.get('artist') and tag_dict.get('artist') != None:
        song_file.tag.artist = tag_dict.get('artist')
    if tag_dict.get('album') and tag_dict.get('album') != None:
        song_file.tag.album = tag_dict.get('album')
    if tag_dict.get('genre') and tag_dict.get('genre') != None:
        song_file.tag.genre = tag_dict.get('genre')
    if tag_dict.get('track') and tag_dict.get('track') != None:
        song_file.tag._setTrackNum(int(tag_dict.get('track')))
    if tag_dict.get('year') and tag_dict.get('year') != None:
        song_file.tag._setReleaseDate(tag_dict.get('year'))
    if tag_dict.get('plays') and tag_dict.get('plays') != None:
        song_file.tag.playcount(int(tag_dict.get('plays')))
    song_file.tag.save(preserve_file_time=True)
    print('success')

def getArtistDir(artist):
    if artist[0].isalpha() == True:
        first_let = artist[0].lower()
    else:
        first_let = '0'
    artist_dir = first_let + os.sep
    return artist_dir

def emptyDB():
    organizer = dict(last_id = 0, ignore_dirs = ['cfg', 'library', 'import'], ids =  {})
    return organizer

def chunkReader(fobj, chunk_size=1024):
    """Generator that reads a file in chunks of bytes"""
    while True:
        chunk = fobj.read(chunk_size)
        if not chunk:
            return
        yield chunk

organizer = openDB()

def checkForDups(file_hash):
    ids = openDB().get('ids')
    for item in ids:
        if ids[item]['uid'] == file_hash:
            print("DUP BITCH")
            return True
    return False

def addToDB(song, hash=sha1):
    print("adding")
    song_path = root_dir + song

    hash_obj = hash()
    for chunk in chunkReader(open(song_path, 'rb')):
        hash_obj.update(chunk)
    file_id = (hash_obj.digest(), os.path.getsize(song_path))
    file_hash = str(file_id[0])

    if checkForDups(file_hash):
        print('sorry')
    else:
        ids = organizer.get('ids')
        tags = getTag(song_path)
        last_id = organizer.get("last_id")
        print(last_id)
        new_id = last_id + 1
        while new_id in ids:
            new_id += 1
        ids[new_id] = tags.copy()
        id_entry = ids[new_id]

        id_entry['uid'] = file_hash
        id_entry['size'] = file_id[1]

        artist = tags.get('artist')
        title = tags.get('title')
        if artist != None and title != None:
            id_entry['file'] = addToLibrary(song, artist, title)
            id_entry['sorted'] = True
        else:
            id_entry['file'] = song
            id_entry['sorted'] = False
        organizer['last_id'] = new_id
        print(organizer.items())
        buildDB(organizer)
    return 'susss'


def getAllIDs():
    id_db = openDB()
    ids = id_db.get('ids')
    id_list = []
    for num in ids:
        id_list.append(num)
    return id_list

def getLibraryIDs():
    id_db = openDB()
    ids = id_db.get('ids')
    id_list = []
    for num in ids:
        if ids[num]['sorted'] == True:
            id_list.append(num)
    return id_list

def getByID(song_id):
    id_db = openDB()
    ids = id_db.get('ids')
    for num in ids:
        if num == str(song_id):
            return ids[num]
    return False
##NEED TO CHOWN ELI:ELI 755 ALL FILES IN MUSIC


player.js

var typingTimer;
var doneTypingInterval = 700;
var musicdb = null;
var querylist = null;
var playlist = null;
var currenti = null;
var ws = null;
var player = document.getElementById('player');

function scrollToLetter(element_id){
    var element = document.getElementById(element_id);
    element.scrollIntoView({behavior:"smooth"});
}
function scrollIntoViewIfNeeded(target) {
    var rect = target.getBoundingClientRect();
    if (rect.bottom > window.innerHeight) {
        target.scrollIntoView({block: "end", behavior: "smooth"});
    }
    if (rect.top < 0) {
        target.scrollIntoView({behavior:"smooth"});
    }
}
function insertAfter(newNode, referenceNode) {
    referenceNode.parentNode.insertBefore(newNode, referenceNode.nextSibling);
}
function clearUndefined(string){
    if (string == null){
        return "\u200b";
    }
    else {
        return string;
    }
}
function emptyUndefined(string){
    if (string == null){
        return "";
    }
    else {
        return string;
    }
}
function getArtistDir(filename){
    var regEx = /^[a-zA-Z\-]+$/;
        if (filename.substring(0,1).toLowerCase().match(regEx)) {
            return filename.substring(0,1).toLowerCase();
        }
        else {
            return '0';
    }
}
function start(){
    ws = new WebSocket("ws://192.168.0.101:8080/websocket");
}
function renewDB(){
    if(!ws || ws.readyState == 3)
        start();
    ws.send("data");
    ws.onopen = function() {};
    ws.onmessage = function (evt){
        musicdb = JSON.parse("[" +evt.data+ "]")[0];
        console.log("DB Received");
        ws.close();
    };
    ws.onclose = function () {
        start();
    };
}

function query(){
    if(!ws || ws.readyState == 3)
        start();
    ws.send("sear" + document.getElementById('search').value);
    ws.onopen = function() {};
    ws.onmessage = function (evt){
        console.log("Search received, clearing songlist");
        insertSongs(JSON.parse("[" +evt.data+ "]")[0]);
        ws.close();
    };
    ws.onclose = function () {
        start();
    };
}

function insertSongs(queryresults){
    var songlist = document.getElementById("songlist");
    var node = songlist;
    while (songlist.hasChildNodes()){
        if (node.hasChildNodes()){
            node = node.lastChild;
        }
        else {
            node = node.parentNode;
            node.removeChild(node.lastChild);
        }
    }
    querylist = Object.keys(queryresults);
    for (var id in queryresults) {
        if (queryresults.hasOwnProperty(id)) {
            var intd = id.slice(1);
            var songbtn = document.createElement("button");
            songbtn.setAttribute("id", id);
            songbtn.setAttribute("class", "btn");
            songbtn.setAttribute('onclick', "startPlaylist(" + intd + ");");
            songbtn.setAttribute('oncontextmenu', "showOps(this); return false;");
            var track = clearUndefined(queryresults[id].track);
            var title = queryresults[id].title;
            if (title == null) {
                title = musicdb[intd];
            }
            var length = clearUndefined(queryresults[id].length);
            var artist = clearUndefined(queryresults[id].artist);
            var album = clearUndefined(queryresults[id].album);
            var year = clearUndefined(queryresults[id].year);
            var bitrate = clearUndefined(queryresults[id].bitrate) + "k";
            var trackdiv = document.createElement("div");
            trackdiv.setAttribute("class", "track");
            if (track != "\u200b"){
                track = track.replace(/^0+/, "");
            }
            trackdiv.textContent = track;
            songbtn.appendChild(trackdiv);
            var titlediv = document.createElement("div");
            titlediv.setAttribute("class", "title");
            titlediv.textContent = title;
            songbtn.appendChild(titlediv);
            var lengthdiv = document.createElement("div");
            lengthdiv.setAttribute("class", "detail");
            lengthdiv.textContent = length;
            songbtn.appendChild(lengthdiv);
            var artistdiv = document.createElement("div");
            artistdiv.setAttribute("class", "artist");
            artistdiv.textContent = artist;
            songbtn.appendChild(artistdiv);
            var albumdiv = document.createElement("div");
            albumdiv.setAttribute("class", "album");
            albumdiv.textContent = album;
            songbtn.appendChild(albumdiv);
            var yeardiv = document.createElement("div");
            yeardiv.setAttribute("class", "detail");
            yeardiv.textContent = year;
            songbtn.appendChild(yeardiv);
            var bitdiv = document.createElement("div");
            bitdiv.setAttribute("class", "detail");
            bitdiv.textContent = bitrate;
            songbtn.appendChild(bitdiv);
            songlist.appendChild(songbtn);
        }
    }
}

function playSong(song_id) {
    var old_song = document.getElementById(playlist[currenti].slice(1));
    if (old_song){
        old_song.setAttribute('class', 'plbtn');
        old_song.children[0].textContent = currenti+1;
    }
    player.pause();
    currenti = playlist.indexOf("_"+song_id);
    var new_song = document.getElementById(song_id);
    new_song.setAttribute('class', 'cplbtn');
    new_song.children[0].innerHTML = "&#9658;";
    scrollIntoViewIfNeeded(new_song);
    player.src = "/static/library/" + getArtistDir(musicdb[song_id]) + "/" + musicdb[song_id];
    player.load();
    player.play();
    updateUI(song_id);
}

function playNext() {
    if (playlist) {
        if (currenti + 1 == playlist.length){
            playSong(playlist[0].slice(1));
        }
        else{
            playSong(playlist[currenti+1].slice(1));
        }
    }
}

function playLast() {
    if (playlist) {
        if (currenti == 0){
            playSong(playlist[0].slice(1));
        }
        else{
            playSong(playlist[currenti-1].slice(1));
        }
    }
}

function startPlaylist(song_id){
    var art = document.getElementById('bounceart');
    if (art){
        document.getElementById('bouncer').removeChild(art);
    }
    var nowplaying = document.getElementById('nowplaying');
    var starti = querylist.indexOf("_"+song_id);
    playlist = querylist.slice(starti);
    starti = 0;
    currenti = starti;
    var node = nowplaying;
    while (nowplaying.hasChildNodes()){
        if (node.hasChildNodes()){
            node = node.lastChild;
        }
        else {
            node = node.parentNode;
            node.removeChild(node.lastChild);
        }
    }
    for (var i=starti, len=playlist.length; i < len; i++) {
        var intd = playlist[i].slice(1);
        var songbtn = document.createElement("button");
        songbtn.setAttribute("id", intd);
        songbtn.setAttribute("class", "plbtn");
        songbtn.setAttribute('onclick', "playSong(" + intd + ");");
        /*songbtn.setAttribute('oncontextmenu', "showOps(this); return false;");*/
        var track = i+1-starti;
        var song_tags = document.getElementById(playlist[i]).children;
        var title = song_tags[1].textContent;
        var artist = song_tags[3].textContent;
        var trackdiv = document.createElement("div");
        trackdiv.setAttribute("class", "pltrack");
        trackdiv.textContent = track;
        songbtn.appendChild(trackdiv);
        var titlediv = document.createElement("div");
        titlediv.setAttribute("class", "pltitle");
        titlediv.textContent = title;
        songbtn.appendChild(titlediv);

        var artistdiv = document.createElement("div");
        artistdiv.setAttribute("class", "plartist");
        artistdiv.textContent = artist;
        songbtn.appendChild(artistdiv);

        nowplaying.appendChild(songbtn);
    }
    playSong(song_id);
}

function insertNext(song_id){
    if (playlist){
        var nowplaying = document.getElementById('nowplaying');
        playlist.splice(currenti+1, 0, "_"+song_id);
        var current_song = document.getElementById(playlist[currenti].slice(1));
        var songbtn = document.createElement("button");
        songbtn.setAttribute("id", song_id);
        songbtn.setAttribute("class", "plbtn");
        songbtn.setAttribute('onclick', "playSong(" + song_id + ");");
        /*songbtn.setAttribute('oncontextmenu', "showOps(this); return false;");*/
        var track = currenti+2;
        var song_tags = document.getElementById("_"+song_id).children;
        var title = song_tags[1].textContent;
        var artist = song_tags[3].textContent;
        var trackdiv = document.createElement("div");
        trackdiv.setAttribute("class", "pltrack");
        trackdiv.textContent = track;
        songbtn.appendChild(trackdiv);
        var titlediv = document.createElement("div");
        titlediv.setAttribute("class", "pltitle");
        titlediv.textContent = title;
        songbtn.appendChild(titlediv);

        var artistdiv = document.createElement("div");
        artistdiv.setAttribute("class", "plartist");
        artistdiv.textContent = artist;
        songbtn.appendChild(artistdiv);
        var songlist = nowplaying.children;
        for (var song in songlist){
            if (songlist.hasOwnProperty(song) && song > currenti) {
                songlist[song].children[0].textContent++;
            }
        }

        insertAfter(songbtn, current_song);
    }
}

function updateUI(id){
    if(!ws || ws.readyState == 3)
        start();
    ws.send("tags" + id);
    ws.onopen = function() {};
    ws.onmessage = function (evt){
        var tag = JSON.parse((evt.data));
        var title = tag.title;
        if (title == null){
            title = tag.file;
        }
        document.getElementById("title").textContent = title;
        document.getElementById("artist").textContent = tag.artist;
        document.getElementById("album").textContent = tag.album;
        var art_path = '/static/cfg/extras/albumart/';
        var image = null;
        image = tag.img;
        if (image) {
            document.getElementById("albumart").setAttribute('src', art_path + image);
        }
        else {
            document.getElementById("albumart").setAttribute('src', '/static/cfg/extras/assets/default-artwork.png');
        }
        document.getElementById("time").textContent = tag.length;
        ws.close();
    };
    ws.onclose = function () {
        start();
    };
}

function clickCheck(term){
    document.getElementById("search").value = term;
    document.getElementById('youtubelink').setAttribute('href', "https://www.youtube.com/results?search_query=" + term);
    console.log("Click search...");
    query();
}

function saveDetails(id) {
    var tag_table = document.getElementById('tag_table');
    var title = tag_table.rows[1].cells[0].textContent;
    var artist = tag_table.rows[1].cells[1].textContent;
    var album = tag_table.rows[1].cells[2].textContent;
    var track = tag_table.rows[1].cells[3].textContent;
    var albumartist = tag_table.rows[3].cells[1].textContent;
    var year = tag_table.rows[3].cells[2].textContent;
    var genre = tag_table.rows[3].cells[3].textContent;
    var artwork = tag_table.rows[5].cells[0].textContent;
    var format_tag = id + "^^^" + title + "^^^" + artist + "^^^" + album + "^^^" + track + "^^^" + albumartist + "^^^" + year + "^^^" + genre + "^^^" + artwork;

    if(!ws || ws.readyState == 3)
        start();
    ws.send("upda" + format_tag);
    document.getElementById("savebtn").innerHTML = "Updating...";
    ws.onopen = function() {};
    ws.onmessage = function (evt){
        musicdb = null;
        ws.close();
        console.log("Reloading DB...");
        checkIfSocketOpen = function()
        {
            if (ws.readyState == 1){
                renewDB();
                console.log("DB Reloaded!");
                clearInterval(loadInterval);
            }
        };
        var loadInterval = setInterval(checkIfSocketOpen, 100);

        checkIfDB = function()
        {
            if (musicdb != null && ws.readyState == 1){
                query();
                console.log("Query Reexicuted!");
                clearInterval(queryInterval);
            }
        };
        var queryInterval = setInterval(checkIfDB, 250);
    };
    ws.onclose = function () {
        start();
    };

}

var file_tag = null;
function showDetails(id) {
    var old_table = document.getElementById('tag_table');
    var closebtn = true;
    if (old_table){
        document.getElementById('detailsbtn').textContent = "Details...";
        closebtn = false;
        songlist.removeChild(old_table);
    }
    if (closebtn){
        document.getElementById('detailsbtn').textContent = "Close";
        var tag_table = document.createElement("table");
        tag_table.id = "tag_table";
        var head = tag_table.createTHead();
        var row1 = head.insertRow(0);
        var cell1 = row1.insertCell(0);
        cell1.innerHTML = "Title";
        var cell2 = row1.insertCell(1);
        cell2.innerHTML = "Artist";
        var cell3 = row1.insertCell(2);
        cell3.innerHTML = "Album";
        var cell4 = row1.insertCell(3);
        cell4.innerHTML = "Track";
        var row2 = tag_table.insertRow(1);
        row2.setAttribute('contentEditable', true);
        var cell5 = row2.insertCell(0);
        cell5.innerHTML = emptyUndefined(file_tag.title);
        var cell6 = row2.insertCell(1);
        cell6.innerHTML = emptyUndefined(file_tag.artist);
        var cell7 = row2.insertCell(2);
        cell7.innerHTML = emptyUndefined(file_tag.album);
        var cell8 = row2.insertCell(3);
        cell8.innerHTML = emptyUndefined(file_tag.track);

        var row3 = head.insertRow(2);
        var cell9 = row3.insertCell(0);
        cell9.innerHTML = "Length";
        var cell10 = row3.insertCell(1);
        cell10.innerHTML = "Album Artist";
        var cell11 = row3.insertCell(2);
        cell11.innerHTML = "Year";
        var cell12 = row3.insertCell(3);
        cell12.innerHTML = "Genre";
        var row4 = tag_table.insertRow(3);
        row4.setAttribute('contentEditable', true);
        var cell13 = row4.insertCell(0);
        cell13.innerHTML = file_tag.length;
        cell13.setAttribute('contentEditable', false);
        var cell14 = row4.insertCell(1);
        cell14.innerHTML = emptyUndefined(file_tag.albumartist);
        var cell15 = row4.insertCell(2);
        cell15.innerHTML = emptyUndefined(file_tag.year);
        var cell16 = row4.insertCell(3);
        cell16.innerHTML = emptyUndefined(file_tag.genre);

        var image = file_tag.img;
        if (image){
            var artcell = row1.insertCell(4);
            artcell.rowSpan = 0;
            artcell.innerHTML = "<img src='/static/cfg/extras/albumart/" + image + "' width=200 height=200>";
        }
        var row5 = head.insertRow(4);
        var cell17 = row5.insertCell(0);
        cell17.innerHTML = "Artwork";
        var row6 = head.insertRow(5);
        row6.setAttribute("contentEditable", true);
        var urlcell = row6.insertCell(0);
        urlcell.colSpan = 0;
        urlcell.innerHTML = "url";
        var row7 = head.insertRow(6);
        var cell18 = row7.insertCell(0);
        cell18.innerHTML = "<br><button id='savebtn' onclick='saveDetails(" + id + ");'>Update</button>";

        insertAfter(tag_table, ops_table);
        scrollIntoViewIfNeeded(tag_table);
    }
}
function showOps(song) {
    var songlist = document.getElementById("songlist");
    var id = song.id.slice(1);
    if(!ws || ws.readyState == 3)
        start();
    ws.send("tags" + id);
    ws.onopen = function() {};
    ws.onmessage = function (evt){
        var old_table = document.getElementById('tag_table');
        if (old_table){
            songlist.removeChild(old_table);
        }
        var old_ops = document.getElementById('ops_table');
        if (old_ops){
            songlist.removeChild(old_ops);
        }

        file_tag = JSON.parse(evt.data);
        var ops_table = document.createElement("table");
        ops_table.id = "ops_table";
        var ophead = ops_table.createTHead();
        var oprow = ophead.insertRow(0);
        var op1 = oprow.insertCell(0);
        op1.innerHTML = "<button id='detailsbtn' onclick='insertNext(" + id + ");'>Play Next</button>";
        var op2 = oprow.insertCell(1);
        var artist_dir = getArtistDir(file_tag.file);
        op2.innerHTML = "<a href='/download/library/" + artist_dir + "/" + file_tag.file + "'  target='_blank'>Download</a>";
        var op3 = oprow.insertCell(2);
        op3.innerHTML = "Lyrics";
        var op4 = oprow.insertCell(3);
        op4.innerHTML = "<button id='detailsbtn' onclick='showDetails(" + id + ");'>Details...</button>";
        insertAfter(ops_table, song);
        ws.close();
    };
    ws.onclose = function () {
        start();
    };
}

function bounceart(path) {
    var art = document.getElementById('bounceart');
    var bouncer = document.getElementById('bouncer');
    var artafter = document.getElementById('artafter');
    if (art){
        bouncer.removeChild(art);
    }
    else{
        art = document.createElement('img');
        art.setAttribute('id', 'bounceart');
        art.setAttribute('src', path);
        art.setAttribute('onclick', 'bounceart(this.src)');
        insertAfter(art, artafter);
    }
}
//Detect keystroke and only execute after the user has finish typing
function delaySearch()
{
    clearTimeout(typingTimer);
        typingTimer = setTimeout(
        function(){
            document.getElementById('youtubelink').setAttribute('href', "https://www.youtube.com/results?search_query=" + document.getElementById('search').value);
            console.log("Executing seach...");
            query();
        },
        doneTypingInterval
    );
    return true;
}

var progressBar = document.querySelector(".progress");
var volumeBar = document.getElementById('volume');
var playbtn = document.getElementById("playpause");

player.addEventListener('progress', function() {
    var bufferedEnd = player.buffered.end(0);
    var duration =  player.duration;
    if (duration > 0) {
        document.getElementById('buffered-amount').style.width = ((bufferedEnd / duration)*100) + "%";
    }
});

player.addEventListener('timeupdate', function() {
    var duration =  player.duration;
    if (duration > 0) {
        document.getElementById('progress-amount').style.width = ((player.currentTime / duration)*100) + "%";
        var numminutes = Math.floor((((player.currentTime % 31536000) % 86400) % 3600) / 60);
        var numseconds = Math.floor((((player.currentTime % 31536000) % 86400) % 3600) % 60);
        if (numseconds < 10) {
            numseconds = '0' + numseconds;
        }
        document.getElementById('progress-time').textContent = numminutes + ":" + numseconds;
    }
});

player.addEventListener('ended', function(){
    playNext();
});

progressBar.addEventListener("click", seek);
function seek(e) {
    var percent = e.offsetX / this.offsetWidth;
    player.currentTime = percent * player.duration;
    progressBar.value = percent * 100;
}


volumeBar.addEventListener("change", setVolume);
function setVolume() {
    player.volume = volumeBar.value;
}

playbtn.addEventListener("click", playPause);
function playPause() {
    player.paused ? player.play() : player.pause();
}

document.body.onkeyup = function(e){
    if(e.keyCode == 32){
        if ((document.activeElement.nodeName == 'INPUT' && document.activeElement.parentNode != document.getElementById('footer')) || (document.activeElement.isContentEditable === true)){}
        else {
            e.preventDefault();
            playPause();
        }
    }
    if (e.keyCode == 27){
        var art = document.getElementById('bounceart');
        if (art){
            document.getElementById('bouncer').removeChild(art);
        }
    }
    if (e.keyCode == 37){
        if ((document.activeElement.nodeName == 'INPUT' && document.activeElement.parentNode != document.getElementById('footer')) || (document.activeElement.isContentEditable === true)){}
        else{
            playLast();
        }
    }
    if (e.keyCode == 39){
        if ((document.activeElement.nodeName == 'INPUT' && document.activeElement.parentNode != document.getElementById('footer')) || (document.activeElement.isContentEditable === true)){}
        else{
            playNext();
        }
    }
};


document.addEventListener("DOMContentLoaded", function() {
    start();

    checkIfSocketOpen = function()
    {
        if (ws.readyState == 1){
            renewDB();
            clearInterval(loadInterval);
        }
    };
    var loadInterval = setInterval(checkIfSocketOpen, 100);

    checkIfDB = function()
    {
        if (musicdb != null && ws.readyState == 1){
            query();
            clearInterval(queryInterval);
        }
    };
    var queryInterval = setInterval(checkIfDB, 250);

});

window.onbeforeunload = function() {
    ws.onclose = function () {};
    ws.close();
    musicdb = null;
    playlist = null;
    ws = null;
    file_tag = null;

    return null;
};

window.addEventListener("beforeunload", function(e) {
    ws.onclose = function () {};
    ws.close();
    e.preventDefault();
    musicdb = null;
    playlist = null;
    ws = null;
    file_tag = null;

    return null;
}, false);
